saq/queue.py

```diff
@@ -394,12 +437,141 @@ async def enqueue(self, job_or_func, **kwargs):
         job.queued = now()
         job.status = Status.QUEUED
 
-        if not await self._enqueue_script(
-            keys=[self._incomplete, job.id, self._queued, job.abort_id],
-            args=[self.serialize(job), job.scheduled],
-            client=self.redis,
-        ):
-            return None
+        for cb in self._before_enqueues.values():
+            await cb(job)
+
+        async with self._enqueue_sem:
+            if not await self._enqueue_script(
+                keys=[self._incomplete, job.id, self._queued, job.abort_id],
+                args=[self.serialize(job), job.scheduled],
+                client=self.redis,
+            ):
+                return None
 
         logger.info("Enqueuing %s", job)
         return job
+
+    async def apply(self, job_or_func, timeout=None, **kwargs):
+        """
+        Enqueue a job and wait for its result.
+
+        If the job is successful, this returns its result.
+        If the job is unsuccessful, this raises a JobError.
+
+        Example::
+            try:
+                assert await queue.apply("add", a=1, b=2) == 3
+            except JobError:
+                print("job failed")
+
+        job_or_func: Same as Queue.enqueue
+        kwargs: Same as Queue.enqueue
+        """
+        results = await self.map(job_or_func, timeout=timeout, iter_kwargs=[kwargs])
+        return results[0]
+
+    async def map(
+        self, job_or_func, iter_kwargs, timeout=None, return_exceptions=False, **kwargs
+    ):
+        """
+        Enqueue multiple jobs and collect all of their results.
+
+        Example::
+            try:
+                assert await queue.map(
+                    "add",
+                    [
+                        {"a": 1, "b": 2},
+                        {"a": 3, "b": 4},
+                    ]
+                ) == [3, 7]
+            except JobError:
+                print("any of the jobs failed")
+
+        job_or_func: Same as Queue.enqueue
+        iter_kwargs: Enqueue a job for each item in this sequence. Each item is the same
+            as kwargs for Queue.enqueue.
+        timeout: Total seconds to wait for all jobs to complete. If None (default) or 0, wait forever.
+        return_exceptions: If False (default), an exception is immediately raised as soon as any jobs
+            fail. Other jobs won't be cancelled and will continue to run.
+            If True, exceptions are treated the same as successful results and aggregated in the result list.
+        kwargs: Default kwargs for all jobs. These will be overridden by those in iter_kwargs.
+        """
+        iter_kwargs = [
+            {"timeout": timeout, "key": kwargs.get("key") or uuid1(), **kwargs, **kw}
```

Comment:
sorry to be picky here, but i don't like that we have to understand to do or uuid1,

wouldn't it be easier to just do Job(key).id

Reply:
`Job` has a required `function` attribute. We can set it to some dummy value, but that also felt ugly.

Reply:
can we have the job iniatilizer use the same helper function and add uuid1() into it as default?

Reply:
yea i'm not sure exactly how to do this

Reply:
Seems sketchy to have implicit default logic in the Job constructor, since that's part of the public API and it might make constructing Jobs more confusing.

We could just add a `get_default_key` function in the job module 🤷 

We could also add a `prepare_job` method 😛

Reply:
ok, yea let's add a job default key